src/workers/parser/getOutOfScopeLocations.js

```diff
@@ -38,26 +40,62 @@ function getLocation(func) {
 }
 
 /**
- * Reduces an array of locations to remove items that are completely enclosed
- * by another location in the array.
+ * Find the nearest location containing the input position and
+ * return new locations without inner locations under that nearest location
+ *
+ * @param locations Notice! The locations MUST be sorted by `sortByStart`
+ *                  so that we can do linear time complexity operation.
  */
-function removeOverlaps(
-  locations: AstLocation | AstLocation[],
-  location: AstLocation
-) {
-  // support reducing without an initializing array
-  if (!Array.isArray(locations)) {
-    locations = [locations];
+function removeInnerLocations(locations: AstLocation[], position: AstPosition) {
+  // First, let's find the nearest position-enclosing function location,
+  // which is to loop locations to find the last location enclosing the position.
+  const newLocs = locations.slice();
+  const parentIndex = findLastIndex(newLocs, loc =>
+    containsPosition(loc, position)
+  );
+  if (parentIndex < 0) {
+    return newLocs;
   }
 
-  const contains =
-    locations.filter(a => containsLocation(a, location)).length > 0;
+  // Second, from the nearest location, loop locations again, stop looping
+  // once seeing the 1st location not enclosed by the nearest location
+  // to find the last inner locations inside the nearest location.
+  const innerStartIndex = parentIndex + 1;
+  let parentLoc = newLocs[parentIndex];
+  let outerBoundaryIndex = findIndex(
+    newLocs,
+    loc => !containsLocation(parentLoc, loc),
+    innerStartIndex
+  );
+  const innerBoundaryIndex =
+    outerBoundaryIndex < 0 ? newLocs.length - 1 : outerBoundaryIndex - 1;
+
+  // Third, remove those inner functions
+  newLocs.splice(innerStartIndex, innerBoundaryIndex - parentIndex);
+  return newLocs;
+}
 
-  if (!contains) {
-    locations.push(location);
+/**
+ * Return an new locations array which excludes
+ * items that are completely enclosed by another location in the input locations
+ *
+ * @param locations Notice! The locations MUST be sorted by `sortByStart`
+ *                  so that we can do linear time complexity operation.
+ */
+function removeOverlaps(locations: AstLocation[]) {
+  if (locations.length == 0) {
```

Comment:
Important to take care of an empty array. Otherwise
```js
// firstParent will be undefined if locations are empty
const firstParent = locations[0]; 
````